File: sample-push-notifications/sample-push-notifications/NotificationsProvider.m
//
//  UserNotificationsHandler.m
//  sample-push-notifications
//
//  Created by Injoit on 3/20/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "NotificationsProvider.h"
#import <UserNotifications/UserNotifications.h>
#import "SVProgressHUD.h"
#import <Quickblox/Quickblox.h>
#import <PushKit/PushKit.h>
#import <CallKit/CallKit.h>

@interface NotificationsProvider() <UNUserNotificationCenterDelegate, PKPushRegistryDelegate>
@property (strong, nonatomic) CXProvider *provider;
@property (strong, nonatomic) CXCallController *callController;
@property (strong, nonatomic) PKPushRegistry *voipRegistry;
@end

@implementation NotificationsProvider

- (instancetype)init {
    self = [super init];
    if (self != nil) {
        
        [self registerForRemoteNotifications];
        
        CXProviderConfiguration *configuration = [self configuration];
        self.provider = [[CXProvider alloc] initWithConfiguration:configuration];
        self.callController = [[CXCallController alloc] initWithQueue:dispatch_get_main_queue()];
        self.voipRegistry = [[PKPushRegistry alloc] initWithQueue:dispatch_get_main_queue()];
        self.voipRegistry.delegate = self;
        self.voipRegistry.desiredPushTypes = [NSSet setWithObject:PKPushTypeVoIP];
    }
    return self;
}

#pragma mark - APNS
- (void)registerForRemoteNotifications {
    // Enable push notifications
    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
    center.delegate = self;
    [center requestAuthorizationWithOptions:(UNAuthorizationOptionSound |
                                             UNAuthorizationOptionAlert |
                                             UNAuthorizationOptionBadge)
                          completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if (!error) {
            [center getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings *settings) {
                if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [[UIApplication sharedApplication] registerForRemoteNotifications];
                    });
                }
            }];
        } else {
            [SVProgressHUD showErrorWithStatus:error.description];
        }
    }];
}

- (NSString *)parseNotification:(UNNotification *)notification {
    NSDictionary *userInfo = notification.request.content.userInfo;
    NSString *message = userInfo[QBMPushMessageApsKey][QBMPushMessageAlertKey];
    if (message) {
        return message;
    }
    return @"Unreadable message";
}

#pragma mark - UNUserNotificationCenterDelegate
- (void)userNotificationCenter:(UNUserNotificationCenter *)center
       willPresentNotification:(UNNotification *)notification
         withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler {
    NSString *apnsMessage = [NSString stringWithFormat:@"APNS: %@", [self parseNotification:notification]];
    [center removeAllDeliveredNotifications];
    if ([self.delegate respondsToSelector:@selector(notificationsProvider:willPresentMessage:)]) {
        [self.delegate notificationsProvider:self willPresentMessage:apnsMessage];
    }
    completionHandler(UNNotificationPresentationOptionAlert | UNNotificationPresentationOptionBadge | UNNotificationPresentationOptionSound);
}

- (void)userNotificationCenter:(UNUserNotificationCenter *)center
didReceiveNotificationResponse:(UNNotificationResponse *)response
         withCompletionHandler:(void(^)(void))completionHandler {
    NSString *apnsMessage = [NSString stringWithFormat:@"APNS: %@", [self parseNotification:response.notification]];
    NSMutableArray *messages = [NSMutableArray arrayWithObject:apnsMessage];
    
    [center getDeliveredNotificationsWithCompletionHandler:^(NSArray<UNNotification *> * _Nonnull notifications) {
        for (UNNotification *notification in notifications) {
            if ([response.notification isEqual:notification]) {
                continue;
            }
            NSString *alertMessage = [NSString stringWithFormat:@"APNS: %@", [self parseNotification:notification]];
            [messages insertObject:alertMessage atIndex:0];
            [center removeAllDeliveredNotifications];
            dispatch_async(dispatch_get_main_queue(), ^{
                if ([self.delegate respondsToSelector:@selector(notificationsProvider:didReceiveMessages:)]) {
                    [self.delegate notificationsProvider:self didReceiveMessages:messages];
                }
            });
        }
    }];
    completionHandler();
}

#pragma mark - VOIP
- (CXProviderConfiguration *)configuration {
    NSString *appName = @"ObjCPushes";
    CXProviderConfiguration *config = [[CXProviderConfiguration alloc] initWithLocalizedName:appName];
    config.supportsVideo = YES;
    config.maximumCallsPerCallGroup = 1;
    config.maximumCallGroups = 1;
    config.supportedHandleTypes = [NSSet setWithObjects:@(CXHandleTypePhoneNumber), nil];
    config.iconTemplateImageData = UIImagePNGRepresentation([UIImage imageNamed:@"CallKitLogo"]);
    config.ringtoneSound = @"ringtone.wav";
    return config;
}

#pragma mark - PKPushRegistryDelegate protocol
- (void)pushRegistry:(PKPushRegistry *)registry didUpdatePushCredentials:(PKPushCredentials *)pushCredentials forType:(PKPushType)type {
    NSString *deviceIdentifier = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    
    QBMSubscription *subscription = [QBMSubscription subscription];
    subscription.notificationChannel = QBMNotificationChannelAPNSVOIP;
    subscription.deviceUDID = deviceIdentifier;
    subscription.deviceToken = [registry pushTokenForType:PKPushTypeVoIP];
    
    [QBRequest createSubscription:subscription successBlock:^(QBResponse *response, NSArray *objects) {
        NSLog(@"[%@] Create Subscription request - Success",  NSStringFromClass([NotificationsProvider class]));
    } errorBlock:^(QBResponse *response) {
        NSLog(@"[%@] Create Subscription request - Error",  NSStringFromClass([NotificationsProvider class]));
    }];
}

- (void)pushRegistry:(PKPushRegistry *)registry didInvalidatePushTokenForType:(PKPushType)type {
    NSString *deviceIdentifier = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    
    [QBRequest subscriptionsWithSuccessBlock:^(QBResponse * _Nonnull response, NSArray<QBMSubscription *> * _Nullable objects) {
        for (QBMSubscription *subscription in objects) {
            if ([subscription.deviceUDID isEqualToString:deviceIdentifier] && subscription.notificationChannel == QBMNotificationChannelAPNSVOIP) {
                [QBRequest deleteSubscriptionWithID:subscription.ID successBlock:^(QBResponse * _Nonnull response) {
                    NSLog(@"[%@] Unregister Subscription request - Success",  NSStringFromClass([NotificationsProvider class]));
                } errorBlock:^(QBResponse * _Nonnull response) {
                    NSLog(@"[%@] Unregister Subscription request - Error",  NSStringFromClass([NotificationsProvider class]));
                }];
                break;
            }
        }
    } errorBlock:^(QBResponse * _Nonnull response) {
        NSLog(@"[%@] Get Subscription request - Error",  NSStringFromClass([NotificationsProvider class]));
    }];
}

- (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(PKPushType)type withCompletionHandler:(void (^)(void))completion {
    UIApplication *application = [UIApplication sharedApplication];
    if (application.applicationState != UIApplicationStateActive) {
        return;
    }
    if (type == PKPushTypeVoIP && application.applicationState == UIApplicationStateActive) {
        
        NSString *message = @"VOIP PUSH from Admin";
        NSString *alertMessage = [payload.dictionaryPayload objectForKey:@"alertMessage"];
        if (alertMessage) {
            message = alertMessage;
        }
        NSString *voipMessage = [NSString stringWithFormat:@"VOIP: %@", message];
        if ([self.delegate respondsToSelector:@selector(notificationsProvider:didReceiveIncomingVOIPPushWithMessage:)]) {
            [self.delegate notificationsProvider:self didReceiveIncomingVOIPPushWithMessage:voipMessage];
        }
    }
    if (completion) {
        completion();
    }
}

@end

File: sample-push-notifications-swift/sample-push-notifications-swift/Components/NotificationsProvider/NotificationsProvider.swift
//
//  NotificationsProvider.swift
//  sample-push-notifications-swift
//
//  Created by Injoit on 3/20/19.
//  Copyright © 2019 QuickBlox. All rights reserved.
//

import Foundation
import UserNotifications
import PushKit
import CallKit
import Quickblox
import SVProgressHUD

enum PushType {
    case apns
    case apnsVoip
}

protocol NotificationsProviderDelegate: class {
    func notificationsProvider(_ notificationsProvider: NotificationsProvider, didReceive messages: [String])
}

class NotificationsProvider: NSObject {
    weak var delegate: NotificationsProviderDelegate?
    private var voipRegistry: PKPushRegistry = PKPushRegistry(queue: DispatchQueue.main)
    private var provider: CXProvider?
    private var callController: CXCallController?
    
    override init() {
        super.init()
        
        registerForRemoteNotifications()
        
        let configuration: CXProviderConfiguration? = self.configuration()
        if let configuration = configuration {
            provider = CXProvider(configuration: configuration)
        }
        callController = CXCallController(queue: DispatchQueue.main)
        
        voipRegistry.delegate = self
        voipRegistry.desiredPushTypes = Set<PKPushType>([.voIP])
    }
}

//MARK: - APNS
//MARK: - UNUserNotificationCenterDelegate
extension NotificationsProvider: UNUserNotificationCenterDelegate {
    private func registerForRemoteNotifications() {
        let center = UNUserNotificationCenter.current()
        center.delegate = self
        center.requestAuthorization(options: [.sound, .alert, .badge], completionHandler: { granted, error in
            if let error = error {
                SVProgressHUD.showError(withStatus: error.localizedDescription)
            }
            center.getNotificationSettings(completionHandler: { settings in
                if settings.authorizationStatus != .authorized {
                    return
                }
                DispatchQueue.main.async(execute: {
                    UIApplication.shared.registerForRemoteNotifications()
                })
            })
        })
    }
    
    private func parseNotification(_ notification: UNNotification) -> String {
        let userInfo = notification.request.content.userInfo
        guard let messageDict = userInfo[QBMPushMessageApsKey] as? [String: AnyObject],
            let message = messageDict[QBMPushMessageAlertKey] as? String else {
                return "APNS: Unreadable message"
        }
        return "APNS: " + message
    }
    
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        delegate?.notificationsProvider(self, didReceive: [parseNotification(notification)])
        center.removeAllDeliveredNotifications()
        completionHandler([.sound, .alert, .badge])
    }
    
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                didReceive response: UNNotificationResponse,
                                withCompletionHandler completionHandler: @escaping () -> Void) {
        var messages = [parseNotification(response.notification)]
        center.getDeliveredNotifications { (notifications) in
            for notification in notifications {
                if notification == response.notification {
                    continue
                }
                messages.insert(self.parseNotification(notification), at: 0)
            }
            center.removeAllDeliveredNotifications()
            DispatchQueue.main.async {
                self.delegate?.notificationsProvider(self, didReceive: messages)
            }
        }
        completionHandler()
    }
}

//MARK: - VOIP
//MARK: - PKPushRegistryDelegate
extension NotificationsProvider: PKPushRegistryDelegate {
    // MARK: - PKPushRegistryDelegate
    private func configuration() -> CXProviderConfiguration? {
           let appName = Bundle.main.infoDictionary?["CFBundleDisplayName"] as? String
           let config = CXProviderConfiguration(localizedName: appName ?? "")
           config.supportsVideo = true
           config.maximumCallsPerCallGroup = 1
           config.maximumCallGroups = 1
           let supportedHandleTypes: Set = [CXHandle.HandleType.generic, CXHandle.HandleType.phoneNumber]
           config.supportedHandleTypes = supportedHandleTypes
           if let image = UIImage(named: "CallKitLogo") {
               config.iconTemplateImageData = image.pngData()
           }
           config.ringtoneSound = "ringtone.wav"
           return config
       }
    
    func pushRegistry(_ registry: PKPushRegistry, didUpdate pushCredentials: PKPushCredentials, for type: PKPushType) {
        guard let voipToken = registry.pushToken(for: .voIP) else {
            return
        }
        guard let deviceIdentifier = UIDevice.current.identifierForVendor?.uuidString else {
            return
        }
        let subscription = QBMSubscription()
        subscription.notificationChannel = .APNSVOIP
        subscription.deviceUDID = deviceIdentifier
        subscription.deviceToken = voipToken
        
        QBRequest.createSubscription(subscription, successBlock: { response, objects in
            debugPrint("[NotificationsProvider] Create VOIP Subscription request - Success")
        }, errorBlock: { response in
            debugPrint("[NotificationsProvider] Create VOIP Subscription request - Error")
        })
    }
    
    func pushRegistry(_ registry: PKPushRegistry, didInvalidatePushTokenFor type: PKPushType) {
        guard let uuidString = UIDevice.current.identifierForVendor?.uuidString else {
            return
        }
        
        QBRequest.subscriptions(successBlock: { (response, subscriptions) in
            guard let subscriptions = subscriptions, subscriptions.isEmpty == false else {
                return
            }
            for subscription in subscriptions {
                if let subscriptionsUIUD = subscription.deviceUDID,
                   subscriptionsUIUD == uuidString,
                   subscription.notificationChannel == .APNSVOIP {
                    QBRequest.deleteSubscription(withID: subscription.id) { (response) in
                        debugPrint("[NotificationsProvider] Unregister VOIP Subscription request - Success")
                    } errorBlock: { (response) in
                        debugPrint("[NotificationsProvider] Unregister VOIP Subscription request - Error")
                    }
                }
            }
        }) { (response) in
            debugPrint("[NotificationsProvider] Subscriptions request - Error")
        }
    }
    
    func pushRegistry(_ registry: PKPushRegistry,
                      didReceiveIncomingPushWith payload: PKPushPayload,
                      for type: PKPushType,
                      completion: @escaping () -> Void) {
        let application = UIApplication.shared
        if application.applicationState != .active {
            return
        }
        if type == .voIP {
            if application.applicationState == .active {
                var message = "VOIP PUSH from Admin"
                if let alertMessage = payload.dictionaryPayload["alertMessage"] as? String {
                    message = alertMessage
                }
                message = "VOIP: " + message
                delegate?.notificationsProvider(self, didReceive: [message])
                completion()
            }
        }
    }
}
